governator-core/src/main/java/com/netflix/governator/conditional/ConditionalBinder.java

```diff
@@ -0,0 +1,406 @@
+     * the qualifier {@code annotation}
+     */
+    public static <T> ConditionalBinder<T> newConditionalBinder(
+        Binder binder, TypeLiteral<T> type, Annotation annotation) {
+        return newRealConditionalBinder(binder, Key.get(type, annotation));
+    }
+
+    /**
+     * Returns a new ConditionalBinder that tracks all candidate instances of {@code type} with 
+     * the qualifier {@code annotation}
+     */
+    public static <T> ConditionalBinder<T> newConditionalBinder(
+        Binder binder, Class<T> type, Annotation annotation) {
+        return newRealConditionalBinder(binder, Key.get(type, annotation));
+    }
+
+    /**
+     * Returns a new ConditionalBinder that tracks all candidate instances of {@code type} with 
+     * the qualifier {@code annotation}
+     */
+    public static <T> ConditionalBinder<T> newConditionalBinder(Binder binder, TypeLiteral<T> type,
+        Class<? extends Annotation> annotationType) {
+        return newRealConditionalBinder(binder, Key.get(type, annotationType));
+    }
+
+    /**
+     * Returns a new ConditionalBinder that tracks all candidate instances of {@code key} where
+     * key may or may not have a qualifier.
+     */
+    public static <T> ConditionalBinder<T> newConditionalBinder(Binder binder, Key<T> key) {
+        return newRealConditionalBinder(binder, key);
+    }    
+  
+    /**
+     * Returns a new ConditionalBinder that tracks all candidate instances of {@code type} with 
+     * the qualifier {@code annotation}
+     */
+    public static <T> ConditionalBinder<T> newConditionalBinder(Binder binder, Class<T> type,
+            Class<? extends Annotation> annotationType) {
+        return newRealConditionalBinder(binder, Key.get(type, annotationType));
+    }
+
+    static <T> ConditionalBinder<T> newRealConditionalBinder(Binder binder, Key<T> key) {
+        if (binder.currentStage().equals(Stage.PRODUCTION)) {
+            throw new RuntimeException("ConditionalBinder may not be used in Stage.PRODUCTION.  Use Stage.DEVELOPMENT.");
+        }
+
+        binder = binder.skipSources(RealConditionalBinder.class, ConditionalBinder.class);
+        RealConditionalBinder<T> result = new RealConditionalBinder<T>(binder, key);
+        binder.install(result);
+        return result;
+    }
+    
+    /**
+     * Returns a binding builder used to add a new conditional candidate for the key. 
+     * Each bound element must have a distinct value. Only the matching candidate will
+     * be instantiated and its provider cached when the key is first injected.
+     *
+     * <p>It is an error to call this method without also calling one of the
+     * {@code to} methods on the returned binding builder.
+     *
+     * <p>Scoping elements independently supported per binding. Use the {@code in} method
+     * to specify a binding scope.
+     */
+    public abstract LinkedBindingBuilder<T> whenMatch(Conditional obj);
```

Comment:
Guice's Matcher has an abstract implementation that includes 'and' and 'or'.  We could follow the same pattern here.